pkg/drivers/docker_driver.go

```diff
@@ -0,0 +1,371 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package drivers
+
+import (
+	"archive/tar"
+	"bufio"
+	"bytes"
+	"fmt"
+	"github.com/sirupsen/logrus"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/container-structure-test/pkg/types/unversioned"
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+type DockerDriver struct {
+	originalImage string
+	currentImage  string
+	cli           docker.Client
+	env           map[string]string
+	save          bool
+}
+
+func NewDockerDriver(args DriverConfig) (Driver, error) {
+	newCli, err := docker.NewClientFromEnv()
+	if err != nil {
+		return nil, err
+	}
+	return &DockerDriver{
+		originalImage: args.Image,
+		currentImage:  args.Image,
+		cli:           *newCli,
+		env:           nil,
+		save:          args.Save,
+	}, nil
+}
+
+func (d *DockerDriver) Destroy() {
+	// since intermediate images are chained, removing the most current
+	// image (that isn't the original) removes all previous ones as well.
+	if d.currentImage != d.originalImage {
+		if err := d.cli.RemoveImage(d.currentImage); err != nil {
+			logrus.Infof("error removing image: %s", err)
+		}
+	}
+}
+
+func (d *DockerDriver) Setup(envVars []unversioned.EnvVar, fullCommands [][]string) {
+	env := d.processEnvVars(envVars)
+	for _, cmd := range fullCommands {
+		d.currentImage = d.runAndCommit(env, cmd)
+	}
+}
+
+func (d *DockerDriver) Teardown(envVars []unversioned.EnvVar, fullCommands [][]string) {
+	// since we create a new driver for each test, skip teardown commands
+	logrus.Debug("Docker driver does not support teardown commands, since each test gets a new driver. Skipping commands.")
+}
+
+func (d *DockerDriver) ProcessCommand(envVars []unversioned.EnvVar, fullCommand []string) (string, string, int) {
+	var env []string
+	for _, envVar := range envVars {
+		env = append(env, envVar.Key+"="+envVar.Value)
+	}
+	stdout, stderr, exitCode := d.exec(env, fullCommand)
+
+	if stdout != "" {
+		logrus.Infof("stdout: %s", stdout)
+	}
+	if stderr != "" {
+		logrus.Infof("stderr: %s", stderr)
+	}
+	return stdout, stderr, exitCode
+}
+
+func retrieveEnv(d *DockerDriver) func(string) string {
+	return func(envVar string) string {
+		var env map[string]string
+		if env == nil {
+			image, err := d.cli.InspectImage(d.currentImage)
+			if err != nil {
+				return ""
+			}
+			// convert env to map for processing
+			env = convertEnvToMap(image.Config.Env)
+		}
+		return env[envVar]
+	}
+}
+
+func (d *DockerDriver) retrieveEnvVar(envVar string) string {
```

Comment:
Can you explain this?  Its a bit of an odd construction

Reply:
sure. essentially the idea behind this is that we only want to retrieve each environment variable once, and cache the result. to do this we can use a closure over the map of environment variables, such that each time we call `retrieveEnvVar` the resulting map is passed to each subsequent call to the same method. this way we're just passing the envVar map around to all calls of `retrieveEnvVar`, without ever having to store it anywhere.